### ISSUE TITLE
azurerm_cdn_frontdoor_firewall_policy custom_rule negation_condition value is ignored

### ISSUE THREAD

**Issue by j-beaven** [hashicorp/terraform-provider-azurerm#25039]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

>1.3.0

### AzureRM Provider Version

>3.45.0

### Affected Resource(s)/Data Source(s)

azurerm_cdn_frontdoor_firewall_policy

### Terraform Configuration Files

```hcl
resource "azurerm_cdn_frontdoor_firewall_policy" "example" {
  name                              = var.name
  resource_group_name               = var.resource_group_name
  sku_name                          = var.sku_name
  enabled                           = var.enabled
  mode                              = var.mode
  redirect_url                      = var.redirect_url
  custom_block_response_status_code = var.custom_block_response_status_code
  custom_block_response_body        = var.custom_block_response_body

  dynamic "custom_rule" {
    for_each = try(var.custom_rule, {})
    content {
      name                           = custom_rule.value.name
      enabled                        = custom_rule.value.enabled
      priority                       = custom_rule.value.priority
      rate_limit_duration_in_minutes = custom_rule.value.rate_limit_duration_in_minutes
      rate_limit_threshold           = custom_rule.value.rate_limit_threshold
      type                           = custom_rule.value.type
      action                         = custom_rule.value.action

      dynamic "match_condition" {
        for_each = try(custom_rule.value.match_conditions, {})
        content {
          match_variable     = match_condition.value.match_variable
          match_values       = match_condition.value.match_values
          operator           = match_condition.value.operator
          selector           = match_condition.value.selector
          negation_condition = match_condition.value.negation_condition
          transforms         = match_condition.value.transforms
        }
      }
    }
  }

  dynamic "managed_rule" {
    for_each = try(var.managed_rule, {})
    content {
      type    = managed_rule.value.type
      version = managed_rule.value.version
      action  = managed_rule.value.action
      dynamic "exclusion" {
        for_each = try(managed_rule.value.exclusions, {})
        content {
          match_variable = exclusion.value.match_variable
          operator       = exclusion.value.operator
          selector       = exclusion.value.selector
        }
      }
      dynamic "override" {
        for_each = try(managed_rule.value.overrides, {})
        content {
          rule_group_name = override.value.rule_group_name
          dynamic "exclusion" {
            for_each = try(override.value.exclusions, {})
            content {
              match_variable = exclusion.value.match_variable
              operator       = exclusion.value.operator
              selector       = exclusion.value.selector
            }
          }
          dynamic "rule" {
            for_each = try(override.value.rules, {})
            content {
              rule_id = rule.value.rule_id
              action  = rule.value.action
              enabled = rule.value.enabled
              dynamic "exclusion" {
                for_each = try(rule.value.exclusion, {})
                content {
                  match_variable = exclusion.value.match_variable
                  operator       = exclusion.value.operator
                  selector       = exclusion.value.selector
                }
              }
            }
          }
        }
      }
    }
  }
}

variable "custom_rule" {
  type = list(object({
      name                           = string
      action                         = string
      enabled                        = optional(bool, true)
      priority                       = number
      type                           = string
      rate_limit_duration_in_minutes = optional(number, 1)
      rate_limit_threshold           = optional(number, 10)
      match_conditions = list(object({
        match_variable   = string
        match_values     = list(string)
        operator         = string
        selector         = optional(string)
        negation_condition = optional(bool)
        transforms       = optional(list(string), [])
      }))
    }))
    default = []
}
```

### Debug Output/Panic Output

```shell
na
```

### Expected Behaviour

I have a Front Door custom rule for waf which has multiple conditions...

Therefore, in the custom rule, we have a condition where traffic to example.com uses a condition so that traffic to "/block_this/but_allow_that/" is allowed.
But then in the same custom rule, we then have a second condition where all other traffic hitting "/block_this/" is blocked.

Using the portal this is achieved by setting the custom rule "Operation" value to either "is" or "is not".

"/block_this/but_allow_that/" = "is not"
"/block_this/" = "is"

We then have a seperate custom rule is used to allow all traffic to example.com.

This works and I want to recreate the same funcitonality using terraform.

Using terraform, as per the [documentation ](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/cdn_frontdoor_firewall_policy) this appears to be achieved using the "negation_condition" which is a bool.

However, when setting the following, the plan always ignores the value true, and always sets the value to false.

So I want...

"/block_this/but_allow_that/" = "is not" = true
"/block_this/" = "is" = false

 {
    name                           = "negation_condition_custom_rule_example"
    enabled                        = false
    priority                       = 90
    rate_limit_duration_in_minutes = 1
    rate_limit_threshold           = 100
    type                           = "MatchRule"
    action                         = "Block"
    match_conditions = [
    {
      match_variable     = "RequestUri"
      operator           = "Contains"
      **negation_condition = true**

      match_values       = {"/block_this/but_allow_that/"]
    },
    {
      match_variable     = "RequestUri"
      operator           = "Contains"
      **negation_condition = false**

      match_values       = ["/block_this/"]
    }]
  },

### Actual Behaviour

      + custom_rule {
          + action                         = "Block"
          + enabled                        = false
          + name                           = "negation_condition_custom_rule_example"
          + priority                       = 90
          + rate_limit_duration_in_minutes = 1
          + rate_limit_threshold           = 100
          + type                           = "MatchRule"

          + match_condition {
              + match_values       = [
                  + "/allow_this/allow_something/",
                ]
              + match_variable     = "RequestUri"
              **+ negation_condition = false**
              + operator           = "Contains"
              + transforms         = []
            }
          + match_condition {
              + match_values       = [
                  + "/block_this/block_something/",
                ]
              + match_variable     = "RequestUri"
              **+ negation_condition = false**
              + operator           = "Contains"
              + transforms         = []
            }
        }

### Steps to Reproduce

terraform plan  & terraform apply

**Comment by alarischu** [comment]
Dear Maintainers,

I’d like to emphasize the importance of addressing the negation_condition issue in the Terraform AzureRM provider. This limitation severely impacts the ability to configure secure Azure Front Door security policies effectively.

As outlined in this discussion, the negation_condition boolean is essential for defining nuanced blocking rules in custom rulesets that ensure traffic is evaluated correctly against managed rules. Here's why this issue is critical:

Use Case:
To ensure robust security, it's crucial that legitimate traffic is evaluated against managed rules for potential intrusion attempts. For example, if we want to allow traffic to https://mydomain.com/foo only from IP 1.2.3.4, we shouldn't use a rule like this:
```
custom_rules = [
{
  name     = "AllowFooForWhitelisted"
  priority = 100
  enabled  = true
  type     = "MatchRule"
  action   = "Allow"
  match_conditions = [
	{
	  match_variable     = "RequestUri"
	  operator           = "BeginsWith"
	  negation_condition = false
	  match_values       = [ "https://mydomain.com/foo" ]
	},
	{
	  match_variable     = "SocketAddr"
	  operator           = "IPMatch"
	  negation_condition = false
	  match_values       = "1.2.3.4"
	}
  ]
}
]
```

and instead, need to use a rule like this:

```
custom_rules = [  
  {  
    name     = "BlockFooForNotWhitelisted"  
    priority = 100  
    enabled  = true  
    type     = "MatchRule"  
    action   = "Block"  
    match_conditions = [  
      {  
        match_variable     = "RequestUri"  
        operator           = "BeginsWith"  
        negation_condition = false  
        match_values       = [ "https://mydomain.com/foo" ]  
      },  
      {  
        match_variable     = "SocketAddr"  
        operator           = "IPMatch"  
        negation_condition = true  
        match_values       = [ "1.2.3.4" ]  
      }  
    ]  
  }  
]  
```
Without honoring the negation_condition, Terraform cannot differentiate between requests matching and not matching the IP address. This makes it impossible to implement such granular rules, leading to insecure configurations.

Impact on Managed Rule Evaluation:
Negation conditions allow for crafting precise blocking rules that exclude specific traffic while ensuring all other traffic is passed to managed rules for evaluation. If negation_condition is ignored, any attempt to apply this logic in Terraform fails, undermining the security strategy.

Current State:
The issue persists in AzureRM provider version 3.116.0, making Terraform ineffective for creating secure Azure Front Door configurations. This greatly limits its utility for users who rely on Terraform for IaC-based security configuration.

In summary, the inability to configure negation_condition makes Terraform unsuitable for secure Azure Front Door policy management, forcing users to resort to manual configurations, defeating the purpose of Infrastructure as Code. I urge the team to prioritize this issue to restore Terraform's effectiveness for Azure Front Door security policies.

Thank you for your time and continued efforts in maintaining this critical provider.

**Comment by rchauvetclara** [comment]
Hello, up for this issue.
I 'm facing the same in my context.

Have we any chance to have a fix ?

Thanks for your help !
